src/name/ip_address.rs

```diff
@@ -295,10 +295,28 @@ pub(super) fn presented_id_matches_constraint(
     let mut name = untrusted::Reader::new(name);
     let mut constraint_address = untrusted::Reader::new(constraint_address);
     let mut constraint_mask = untrusted::Reader::new(constraint_mask);
+    let mut seen_zero_bit = false;
+
     loop {
         let name_byte = name.read_byte().unwrap();
         let constraint_address_byte = constraint_address.read_byte().unwrap();
         let constraint_mask_byte = constraint_mask.read_byte().unwrap();
+
+        let leading = constraint_mask_byte.leading_ones();
+        let trailing = constraint_mask_byte.trailing_zeros();
+
+        if leading + trailing != 8 {
+            return Err(Error::InvalidNetworkMaskConstraint);
+        }
+
+        if seen_zero_bit && leading > 0 {
+            return Err(Error::InvalidNetworkMaskConstraint);
+        }
+
+        if leading != 8 {
+            seen_zero_bit = true;
+        }
```

Comment:
I think this could do with some prose in comments on what the goal here is. In particular, I don't think I understand why we require that `leading + trailing == 8`.

Reply:
will add some commentary here. it's working off the fact that a valid mask is a sequence of 1 bits followed by 0 bits; on a octet level this is detected by leading_ones + trailing_zeroes summing to 8 in the case of a valid mask, but never in the case of an invalid one. eg:

```
    for j in 0u16..256u16 {
        let i = j as u8;
        if i.leading_ones() + i.trailing_zeros() == 8 {
            println!("{:08b}", i);
        }
    }
```
prints
```
00000000
10000000
11000000
11100000
11110000
11111000
11111100
11111110
11111111
```